lib/jsdom/living/nodes/HTMLDialogElement-impl.js

```diff
@@ -1,8 +1,123 @@
 const HTMLElementImpl = require("./HTMLElement-impl").implementation;
+const { fireAnEvent } = require("../helpers/events");
+const { runAutoFocus, outsideUnopenedDialog } = require("../helpers/focusing");
+
+const ALREADY_OPEN = "Failed to execute 'showModal' on 'HTMLDialogElement': " +
+                    "The element already has an 'open' attribute, and therefore cannot be opened modally.";
+
+const NOT_CONNECTED = "Failed to execute 'showModal' on 'HTMLDialogElement': " +
+                      "The element is not in a Document.";
+
+class HTMLDialogElementImpl extends HTMLElementImpl {
+  constructor(globalObject, args, privateData) {
+    super(globalObject, args, privateData);
+    this._isModal = false;
+    this._returnValue = "";
+    this._previousFocusElement = null;
+  }
+
+  get open() {
+    return this.getAttributeNS(null, "open") !== null;
+  }
+
+  get returnValue() {
+    return this._returnValue;
+  }
+
+  set returnValue(returnValue) {
+    this._returnValue = returnValue.toString();
+  }
+
+  show() {
+    if (this.open) {
+      return;
+    }
+
+    this.setAttributeNS(null, "open", "");
+
+    if (this._previousFocusElement) {
+      this._previousFocusElement.focus();
+    }
+
+    this._dialogFocusSteps();
+  }
+
+  showModal() {
+    if (this.open) {
+      throw DOMException.create(this._globalObject, [ALREADY_OPEN, "InvalidStateError"]);
+    }
+
+    if (!this.isConnected) {
+      throw DOMException.create(this._globalObject, [NOT_CONNECTED, "InvalidStateError"]);
+    }
+
+    this.setAttributeNS(null, "open", "");
+    this._isModal = true;
+
+    this._ownerDocument._modalBlock(this);
+
+    if (this._previousFocusElement) {
+      this._previousFocusElement.focus();
+    }
+
+    this._dialogFocusSteps();
+  }
+
+  _dialogFocusSteps() {
+    if (this.inert) {
+      return;
+    }
+
+    if (runAutoFocus(this)) {
+      return;
+    }
+
+    const delegates = ["button", "input", "select", "textarea"]
+      .map(delegate => `${delegate}:not(:disabled):not([hidden])`);
+
+    const focusDelegates = this.querySelectorAll(
+      `${delegates.join(", ")}, [href], [tabindex]:not([tabindex="-1"]):not(dialog)`
+    );
+
+    const focusDelegate = [...focusDelegates].find(outsideUnopenedDialog);
+
+    if (focusDelegate) {
+      this._previousFocusElement = focusDelegate;
+      focusDelegate.focus();
+      return;
+    }
+
+    const focusable = this.getAttributeNS(null, "tabindex");
+    if (focusable && Number(focusable) > -1) {
+      this._previousFocusElement = this;
+      this.focus();
+    }
+  }
+
+  close(result) {
+    if (!this.open) {
+      return;
+    }
+
+    this.removeAttributeNS(null, "open");
+    this._isModal = false;
+
+    if (result !== undefined) {
+      this._returnValue = result.toString();
+    }
+
+    this._ownerDocument._modalUnblock(this);
+
+    if (this._previousFocusElement) {
+      this._previousFocusElement.blur();
```

Comment:
Step 6.3 of https://html.spec.whatwg.org/#close-the-dialog *focuses* the previous focused element; it doesn't blur it.

Also it's important to clear it before focusing it in case the focus event handler causes something to get messed up.